sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -388,3 +390,36 @@ object PhysicalWindow {
     case _ => None
   }
 }
+
+object ExtractSingleColumnNullAwareAntiJoin extends JoinSelectionHelper with PredicateHelper {
+
+  // SingleColumn NullAwareAntiJoin
+  // streamedSideKeys, buildSideKeys
+  // currently these two return Seq[Expression] should have only one element
+  private type ReturnType = (Seq[Expression], Seq[Expression])
+
+  /**
+   * See. [SPARK-32290]
+   * LeftAnti(condition: Or(EqualTo(a=b), IsNull(EqualTo(a=b)))
+   * will almost certainly be planned as a Broadcast Nested Loop join,
+   * which is very time consuming because it's an O(M*N) calculation.
+   * But if it's a single column case, and buildSide data is small enough,
+   * O(M*N) calculation could be optimized into O(M) using hash lookup instead of loop lookup.
+   */
+  def unapply(join: Join): Option[ReturnType] = join match {
+    case Join(left, right, LeftAnti,
+      Some(Or(equalTo @ EqualTo(leftAttr: AttributeReference, rightAttr: AttributeReference),
+        IsNull(EqualTo(tmpLeft: AttributeReference, tmpRight: AttributeReference)))), _)
+        if SQLConf.get.nullAwareAntiJoinOptimizeEnabled &&
+          leftAttr.semanticEquals(tmpLeft) && rightAttr.semanticEquals(tmpRight) &&
+          canBroadcastBySize(right, SQLConf.get) &&
+          right.output.length == 1 =>
```

Comment:
why do we need this condition? I think we only need to look at the number of join keys.

Reply:
ok, will do the update.